utility/src/main/java/org/oppia/util/parser/LiTagHandler.kt

```diff
@@ -1,37 +1,167 @@
+        lists.pop()
+      }
+      LI_TAG -> if (opening) { // handle <li>
+        lists.peek().openItem(output)
+      } else { // handle </li>
+        lists.peek().closeItem(output, indentation = lists.size - 1)
+      }
+    }
+  }
+
   /**
-   * Helper marker class. Based on [Html.fromHtml] implementation.
+   * Handler for <li> tags. Subclasses set the bullet appearance.
    */
-  class Bullet
+  private interface ListTag {
 
-  override fun handleTag(opening: Boolean, tag: String, output: Editable, xmlReader: XMLReader) {
-    if (tag == "li" && opening) {
-      output.setSpan(Bullet(), output.length, output.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE)
-    }
-    if (tag == "li" && !opening) {
-      output.append("<br>")
-      val lastMark = output.getSpans(0, output.length, Bullet::class.java).lastOrNull()
-      lastMark?.let {
-        val start = output.getSpanStart(it)
-        output.removeSpan(it)
-        if (start != output.length) {
-          output.setSpan(BulletSpan(), start, output.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
-        }
+    /**
+     * Called when an opening <li> tag is encountered.
+     *
+     * Inserts an invisible [Mark] span that doesn't do any styling.
+     * Instead, [closeItem] will later find the location of this span so it knows where the opening tag was.
+     */
+    fun openItem(text: Editable)
+
+    /**
+     * Called when a closing </li> tag is encountered.
+     *
+     * Pops out the invisible [Mark] span and uses it to get the opening tag location.
+     * Then, sets a [LeadingMarginSpan] from the opening tag position to closing tag position.
+     */
+    fun closeItem(text: Editable, indentation: Int)
+  }
+
+  /**
+   * Subclass of [ListTag] for unordered lists.
+   */
+  private inner class Ul : ListTag {
+
+    override fun openItem(text: Editable) {
+      appendNewLine(text)
+      start(text, BulletListItem())
+    }
+
+    override fun closeItem(text: Editable, indentation: Int) {
+      appendNewLine(text)
+
+      getLast<BulletListItem>(text)?.let { mark ->
+        setSpanFromMark(text, mark, CustomBulletSpan(context, GAP_WIDTH, indentation, "•"))
+      }
+    }
+  }
+
+  /**
+   * Subclass of [ListTag] for ordered lists.
+   */
+  private inner class Ol : ListTag {
+
+    private var index = 1
+
+    override fun openItem(text: Editable) {
+      appendNewLine(text)
+      start(text, NumberListItem(index))
+      index++
+    }
+
+    override fun closeItem(text: Editable, indentation: Int) {
+      appendNewLine(text)
+      getLast<NumberListItem>(text)?.let { mark ->
+        setSpanFromMark(text, mark, CustomBulletSpan(context, GAP_WIDTH, indentation, "${mark.number}."))
       }
     }
   }
+
+  /**
+   * These static methods are based on the Android Html class source code.
+   * https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/text/Html.java
+   */
+  companion object {
+
+    private const val GAP_WIDTH = 80
```

Comment:
I think this value should come from `dimens.xml` file so that we can support various density screens.

Reply:
Done.